Add back nested templates for ARM best practices review

hosungsmsft opened the pull request:
PR on a separate branch to collaborate better on updating the templates for ARM best practices. This won't/shouldn't  be completed (no effect on master), but should be applied to master separately.

bmoore-msft commented on nested/blobStorageAccount.json:
```diff
@@ -0,0 +1,58 @@
+                },
+                "supportsHttpsTrafficOnly": true
+            },
+            "type": "Microsoft.Storage/storageAccounts"
```
move this to the top of the resource definition - need to do it for all resources in all files

hosungsmsft replied:
Done

bmoore-msft commented on nested/blobStorageAccount.json:
```diff
@@ -0,0 +1,58 @@
+        "documentation1": "This sub-template creates a blob storage account.  It expects certain values in the 'common' datastructure.",
+        "documentation2": " blobStorageAccountName - name of storage account",
+        "documentation3": " blobStorageAccountType - type of storage account",
+        "scriptUri": "[concat(parameters('moodleCommon').ScriptLocation,parameters('moodleCommon').moodleInstallScriptFilename)]",
```
remove - not used

hosungsmsft replied:
Done on another separate commit

bmoore-msft commented on nested/blobStorageAccount.json:
```diff
@@ -0,0 +1,58 @@
+        "documentation2": " blobStorageAccountName - name of storage account",
+        "documentation3": " blobStorageAccountType - type of storage account",
+        "scriptUri": "[concat(parameters('moodleCommon').ScriptLocation,parameters('moodleCommon').moodleInstallScriptFilename)]",
+        "blobStorageAccountId": "[concat(resourceGroup().id,'/providers/Microsoft.Storage/storageAccounts/', parameters('moodleCommon').blobStorageAccountName)]"
```
use resourceId instead of concat

hosungsmsft replied:
Done

bmoore-msft commented on nested/controller.json:
```diff
@@ -0,0 +1,234 @@
+            "metadata": {
+                "description": "Key used to create a storage container"
+            },
+            "type": "string"
```
secureString

hosungsmsft replied:
Done

hosungsmsft replied:
More episodes... I realized that the keys don't have to be passed around, so I got rid of the outputs and params. Initially I put listKeys() in variables, which is not allowed by ARM. After all, I substituted the whole expressions (including listKeys()) for variables, so now there's no more string secrets being passed around...

bmoore-msft commented on nested/blobStorageAccount.json:
```diff
@@ -0,0 +1,58 @@
+    "outputs": {
+        "blobStorageAccountKey": {
+            "value": "[listKeys(variables('blobStorageAccountId'), '2017-06-01').keys[0].value]",
+            "type": "string"
```
secureString

hosungsmsft replied:
Done

hosungsmsft replied:
Ditto (removed the string secrets outputs/params all together).

bmoore-msft commented on nested/controller.json:
```diff
@@ -0,0 +1,234 @@
+    },
+    "resources": [
+        {
+            "apiVersion": "2015-06-15",
```
update apiVersion - do this through all the files, make sure none of the apiVersions are more than 1 year old

hosungsmsft replied:
Done

bmoore-msft commented on nested/controller.json:
```diff
@@ -0,0 +1,234 @@
+        "nicRef": "[resourceId('Microsoft.Network/networkInterfaces', parameters('moodleCommon').jboxNicName)]",
+        "nsgRef": "[resourceId('Microsoft.Network/networkSecurityGroups', parameters('moodleCommon').jboxNsgName)]",
+        "pipRef": "[resourceId('Microsoft.Network/publicIPAddresses', parameters('moodleCommon').jboxPipName)]",
+        "subnetWebRef": "[concat(resourceId('Microsoft.Network/virtualNetworks', parameters('moodleCommon').vnetName), '/subnets/',parameters('moodleCommon').subnetWeb)]"
```
simplify - remove concat, just use resourceId

hosungsmsft replied:
Done

bmoore-msft commented on nested/diskSelection.json:
```diff
@@ -0,0 +1,92 @@
+                }
+            }
+        ],
+        "documentation01": "This sub-template provides a way to return a flexible number of drive configurations",
```
use a property copy loop for this instead of the nested deployments, see: https://github.com/Azure/azure-quickstart-templates/blob/master/101-vm-windows-copy-datadisks/azuredeploy.json#L150-L160

hosungsmsft replied:
This is great. Modified as such and removed this nested template.

hosungsmsft replied:
I was excited to know about this and try it, but unfortunately, it doesn't work with our current template structure. ARM doesn't allow 0 copy count, but we are depending on an empty dataDisks array, so a property copy loop can't be used in that case (in the controller.json). So I had to revive this nested template, after so many trials and errors. I wish ARM supported 0 copy count...

hosungsmsft replied:
Giving the take() a try, which seems to accept 0 as the count. Quite inconsistent...

bmoore-msft commented on nested/elastic.json:
```diff
@@ -0,0 +1,368 @@
+                    ]
+                },
+                "osProfile": {
+                    "adminPassword": "[parameters('moodleCommon').sshPassword]",
```
why are we mixing password and ssh auth throughout the app?

hosungsmsft replied:
Good catch and I can only guess this might be an oversight from the original creator. Unified to ssh pubkey auth.

bmoore-msft commented on nested/elastic.json:
```diff
@@ -0,0 +1,368 @@
+        "nicRef1": "[resourceId('Microsoft.Network/networkInterfaces', parameters('moodleCommon').elasticNicName1)]",
+        "nicRef2": "[resourceId('Microsoft.Network/networkInterfaces', parameters('moodleCommon').elasticNicName2)]",
+        "nicRef3": "[resourceId('Microsoft.Network/networkInterfaces', parameters('moodleCommon').elasticNicName3)]",
+        "subnetElasticRef": "[concat(resourceId('Microsoft.Network/virtualNetworks', parameters('moodleCommon').vnetName), '/subnets/',parameters('moodleCommon').subnetElastic)]"
```
just use resourceId

hosungsmsft replied:
Done

bmoore-msft commented on nested/glustervm.json:
```diff
@@ -0,0 +1,212 @@
+            "type": "Microsoft.Network/networkInterfaces"
+        },
+        {
+            "apiVersion": "[parameters('moodleCommon').computeApi]",
```
use literal values for apiVersions

hosungsmsft replied:
Done

bmoore-msft commented on nested/glustervm.json:
```diff
@@ -0,0 +1,212 @@
+            "tags": {
+                "displayName": "Gluster VM Data Disk"
+            },
+            "type": "Microsoft.Compute/disks"
```
remove this and just use implicit disks

hosungsmsft replied:
Thanks for the great info. Modified that way.

bmoore-msft commented on nested/glustervm.json:
```diff
@@ -0,0 +1,212 @@
+        "documentation20": " glusterDiskSize         - Size per disk for gluster",
+        "nicName": "[concat(variables('vmName'),'-nic')]",
+        "nicRef": "[resourceId('Microsoft.Network/networkInterfaces', variables('nicName'))]",
+        "subnetSanRef": "[concat(resourceId('Microsoft.Network/virtualNetworks', parameters('moodleCommon').vnetName), '/subnets/',parameters('moodleCommon').subnetSan)]",
```
just use resourceID

hosungsmsft replied:
Done

bmoore-msft commented on nested/controllerconfig.json:
```diff
@@ -0,0 +1,65 @@
+        "documentation10": "    moodleDbUser                - database user for moodle",
+        "documentation11": "    moodleDbPass                - database password for moodleDbUser",
+        "documentation12": "    moodleAdminPass             - password for moodle admin user",
+        "scriptUri": "[concat(parameters('moodleCommon').ScriptLocation,parameters('moodleCommon').moodleInstallScriptFilename)]"
```
See [BP guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md#samples-that-contain-extra-artifacts-custom-scripts-nested-templates-etc) for how to handle artifacts

hosungsmsft replied:
I assume you meant the _artifactsLocationSasToken practice, which was done on a separate commit after this PR was issued. Marked as done as such.

hosungsmsft commented:
@bmoore-msft - Thanks a lot again for your detailed reviews. I addressed all your comments and updated this PR on subsequent commits. However, there were many other commits that were done on master in the meantime and that had to be merged here, so I reset the base of this PR to that point (after merging with master). It'd be great if you do another review to confirm my changes are as you expected. I suppose you may not need to do another round of review and @rgardler may just discuss with you in person about how to proceed.

@rgardler - You may want to review the changes here that address Brian's comments on ARM best practices. I had to do many trials and errors to get all right...

hosungsmsft commented:
Unfortunately, changing the PR base doesn't seem to work for hiding other unrelated changes (that were included from merging master)... Subsequent reviews may not be very straightforward and I understand that this PR can be separately handled.

hosungsmsft commented:
Closing this PR, as this is now moved to the azure-quickstart-templates repo as a [PR](https://github.com/Azure/azure-quickstart-templates/pull/4317) there. The changes in this PR are merged to another branch (hs-loadtest), which will be merged to master laster.
